dataflow/src/domain/mod.rs

```diff
@@ -1507,6 +1516,71 @@ impl Domain {
         }
     }
 
+    fn handle_recovery(&mut self) {
+        let node_info: Vec<_> = self.nodes
+            .iter()
+            .map(|(index, node)| {
+                let n = node.borrow();
+                (index.clone(), n.global_addr(), n.is_transactional())
+            })
+            .collect();
+
+        for (local_addr, global_addr, is_transactional) in node_info {
+            let checktable = self.transaction_state.get_checktable();
+            let path = self.persistence_parameters.log_path(
+                &local_addr,
+                self.index,
+                self.shard.unwrap_or(0),
+            );
+
+            let file = match File::open(&path) {
+                Ok(f) => f,
+                Err(ref e) if e.kind() == ErrorKind::NotFound => continue,
+                Err(e) => panic!("Could not open log file {:?}: {}", path, e),
+            };
+
+            BufReader::new(file)
+                .lines()
+                .filter_map(|line| {
+                    let line = line.unwrap();
+                    let entries: Result<Vec<Records>, _> = serde_json::from_str(&line);
+                    entries.ok()
+                })
+                .flat_map(|r| r)
+                // Merge packets into batches of RECOVERY_BATCH_SIZE:
+                .chunks(RECOVERY_BATCH_SIZE)
+                .into_iter()
+                .map(|chunk| chunk.fold(Records::default(), |mut acc, ref mut data| {
```

Comment:
OOI, why is this step required? AIUI, the original `Vec<Records>` exists because each line can contain several records (one originally-written batch per line?), so the real type is `Vec<Vec<Vec<DataType>>>`. The `flat_map` removes the outermost layer, so we now basically have a single, large `Vec<Vec<DataType>>`, which we then chunk into batches. This turns into a new iterator over these batched chunks (iterator over type `&[Vec<DataType>]`, within which each chunk is concerted into a single `Records` structure (= `Vec<Vec<DataType>>`). Is that correct?

(In other words, the extra `map`/`fold` is required to deal with the slice type and get back to `Records`.)

Reply:
After taking a second look at this I'm pretty sure that the original implementation is actually wrong, thanks for asking about it.

Since it only called `flat_map` once, it would essentially chunk based on `Records` - and not `Record`:

```rust
.filter_map(|line| {
    let line = line.unwrap();
    let entries: Result<Vec<Records>, _> = serde_json::from_str(&line);
    entries.ok()
})
// At this point we have an iterator over `Vec<Records>` (lines in the log)
.flat_map(|r| r)
// We've now gone from an iterator over `Vec<Records>` to an iterator over `Records`
.chunks(...)
// Each chunk now has a list of `Records`, which are flattened into one by the `.fold`.
```

It should be fixed in https://github.com/mit-pdos/distributary/pull/37/commits/cc996f285aa5ce57c9e2ed62729295cc18c60080 - which adds a second `flat_map`, and collects each chunk into a `Records` before passing it on to the packet.